File: Easy Puzzles/Crop-Circles/Crop-Circles.rb
def distance(x1, y1, x2, y2)
    return Math.sqrt( (y2-y1)**2 + (x2-x1)**2 )
end

instructions = gets.chomp.split.map{ |ins| ins.scan(/(PLANT|PLANTMOW)*(.)(.)(\d+)/)[0] }
grid = Array.new(25) { Array.new(19){?x} }

instructions.each do |extra, x, y, d|
    col = x.ord-97
    row = y.ord-97
    rad = d.to_f/2.0

    for y in 0...25
        for x in 0...19
            dis = distance(x, y, col, row)
            if rad >= dis
                if extra.nil?
                    grid[y][x] = ' '
                elsif extra == "PLANT"
                    grid[y][x] = 'x'
                elsif extra == "PLANTMOW"
                    grid[y][x] = grid[y][x] == ?x ? ' ' : ?x
                end
            end
        end
    end
end
grid.map! {_1.join.gsub(?x,'{}').gsub(' ', '  ')}
puts grid

File: Easy Puzzles/Crop-Circles/Crop-Circles.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

double distance(x1, y1, x2, y2)
{
    return sqrt( pow(y2-y1, 2) + pow(x2-x1, 2) );
}

int main()
{
    char instructions[501];
    scanf("%[^\n]", instructions);

    char* field[25];
    // fill with TRUE values
    char* str = "\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01";
    for (int i = 0; i < 25; ++i)
        field[i] = strdup(str);

    char* ins = strtok(instructions, " ");

    int col, row, extra;
    double radius;

    while (ins)
    {
        if (*ins != 'P') // regular instruction
            extra = 0;
        else if (ins[5] == 'M') // PRINTMOW
        {
            ins += 8;
            extra = 1;
        }
        else // PRINT
        {
            ins += 5;
            extra = 2;
        }

        col = *ins - 97;
        row = ins[1] - 97;
        radius = atoi(ins + 2)/2.f;

        for (int y = 0; y < 25; ++y)
        {
            for (int x = 0; x < 19; ++x)
            {
                if (radius >= distance(x, y, col, row))
                {
                    if (!extra)
                        field[y][x] = false;
                    else if (extra == 1)
                        field[y][x] ^= true;
                    else
                        field[y][x] = true;
                }
            }
        }

        ins = strtok(NULL, " ");
    }

    for (int y = 0; y < 25; ++y)
    {
        for (int x = 0; x < 19; ++x)
            printf("%s", field[y][x] ? "{}" : "  ");

        printf("\n");
        free(field[y]);
    }

    return 0;
}
